src/services/download-service.ts

```diff
@@ -12,123 +13,153 @@ class DownloadService {
-      return '';
+  async getDownloadMetadata(): Promise<DownloadMetadata> {
+    if (!(this.isPlatformSupported() && this.isEntrySupported())) {
+      return null;
     }
 
-    const {ks} = playerConfig.session;
-    const {id} = playerConfig.sources;
-    const {partnerId} = provider;
-    const {cdnUrl} = provider.env;
+    const metadata: DownloadMetadata = {
+      fileName: '',
+      captions: [],
+      flavors: [],
+      attachments: [],
+      imageDownloadUrl: ''
+    };
+
+    const assets = await this.getKalturaAssets();
+    Object.assign(metadata, assets);
+    metadata!.imageDownloadUrl = await this.handleImageDownload();
+    const downloadUrls: Map<string, string> = await this.getDownloadUrls(metadata);
+
+    // assign the download urls to the assets by id
+    metadata?.captions.forEach(captionAsset => (captionAsset.downloadUrl = downloadUrls.get(captionAsset.id)!));
+    metadata?.flavors.forEach(flavorAsset => (flavorAsset.downloadUrl = downloadUrls.get(flavorAsset.id)!));
+    metadata?.attachments.forEach(attachmentAsset => (attachmentAsset.downloadUrl = downloadUrls.get(attachmentAsset.id)!));
+
+    metadata!.fileName = this.getFilename(metadata);
+    return metadata;
+  }
 
-    const partnerPart = `/p/${partnerId}`;
-    const entryIdPart = `/entryId/${id}`;
-    const flavorParamIdPart = flavorParamId !== null ? `/flavorParamId/${flavorParamId}` : '';
-    const flavorIdPart = flavorId !== null ? `/flavorId/${flavorId}` : '';
-    const ksPart = ks ? `/ks/${ks}` : '';
-    const protocolPart = `/protocol/${location.protocol.split(':')[0]}`;
+  private async getDownloadUrls(metadata: DownloadMetadata): Promise<Map<string, string>> {
+    const ks = this.player.config.session?.ks || '';
+    let urls = new Map<string, string>();
+
+    const data = await this.player.provider.doRequest(
+      [{loader: DownloadUrlLoader, params: {flavors: metadata?.flavors, captions: metadata?.captions, attachments: metadata?.attachments}}],
+      ks
+    );
+    if (data && data.has(DownloadUrlLoader.id)) {
+      const urlsLoader = data.get(DownloadUrlLoader.id);
+      urls = urlsLoader?.response?.urls;
+    }
 
-    return `${cdnUrl}${partnerPart}/playManifest${entryIdPart}${protocolPart}${ksPart}${flavorParamIdPart}${flavorIdPart}/format/download`;
+    return urls;
   }
 
-  async getDownloadMetadata(config: DownloadConfig): Promise<DownloadMetadata> {
-    if (!(this.isPlatformSupported() && this.isEntrySupported())) {
-      return null;
+  private async handleImageDownload(): Promise<string> {
+    if (!this.player.isImage()) return '';
+    let imageRequestUrl = this.player.sources.downloadUrl;
+    if (!imageRequestUrl) {
+      return await this.getImageDownloadMetadata();
     }
-
-    const requestUrl = this.getRequestUrl(config);
-    if (!requestUrl) {
-      return null;
+    if (!imageRequestUrl.includes('/ks/')) {
```

Comment:
we already have a code in the player that appends ks to thumbnails, maybe it can be reused